pkg/postage/batchstore/reserve.go

```diff
@@ -0,0 +1,230 @@
+// i.e.  chunk of  the batch in bins [0 upto PO] (closed  interval) are unreserved
+func (s *store) unreserve(batchID []byte, radius uint8) error {
+	return s.unreserveFunc(batchID, radius)
+}
+
+func (s *store) purgeExpired() error {
+	var toDelete [][]byte
+	err := s.store.Iterate(valueKeyPrefix, func(key, _ []byte) (bool, error) {
+		b, err := s.Get(valueKeyToID(key))
+		if err != nil {
+			return true, err
+		}
+		if b.Value.Cmp(s.rs.Core) >= 0 {
+			return true, nil
+		}
+		err = s.unreserve(b.ID, swarm.MaxPO)
+		if err != nil {
+			return true, err
+		}
+		// if batch has no value then delete it
+		if b.Value.Cmp(s.cs.Total) <= 0 {
+			toDelete = append(toDelete, b.ID)
+		} else {
+			b.Radius = swarm.MaxPO
+			if err := s.store.Put(batchKey(b.ID), b); err != nil {
+				return true, err
+			}
+		}
+
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+	return s.delete(toDelete...)
+}
+
+type change struct {
+	old, new int
+}
+
+func newChange(oldv, newv, lower, higher *big.Int) *change {
+	return &change{
+		old: cmp(oldv, lower, higher),
+		new: cmp(newv, lower, higher),
+	}
+}
+
+func cmp(x, lower, higher *big.Int) int {
+	if x.Cmp(lower) < 0 || x.Cmp(big.NewInt(0)) == 0 {
+		return -1
+	}
+	if x.Cmp(higher) < 0 {
+		return 0
+	}
+	return 1
+}
+
+func (ch *change) changed() bool {
+	return ch.old != ch.new
+}
+
+func (ch *change) increased() bool {
+	return ch.old < ch.new
+}
+
+func (ch *change) double() bool {
+	return ch.old+ch.new == 0
+}
+
+func (ch *change) out() bool {
+	return ch.new < 0
+}
+
+// updateReserveState manages what chunks of which batch are allocated to the reserve
+func (s *store) updateValueChange(b *postage.Batch, oldDepth uint8, oldValue *big.Int) error {
+	newValue := b.Value
+	newDepth := b.Depth
+	defer func() {
+		if s.rs.Capacity > 0 && s.rs.Core.Cmp(newValue) > 0 {
+			s.rs.Core.Set(newValue)
+		}
+	}()
+	ch := newChange(oldValue, newValue, s.rs.Core, s.rs.Edge)
+	// if values unchanged then limits dont change
+	if !ch.changed() {
+		return nil
+	}
+	half := exp2(newDepth - s.rs.Depth - 1)
+	// if value drops out then available capacity is increased with the batch's size within neighbourhood
+	if ch.out() {
+		s.rs.Capacity += half
+		if ch.double() {
+			s.rs.Capacity += half
+		}
+		// the chunks of this batch are unreserved in all po bins down to MaxPO
+		return s.unreserve(b.ID, swarm.MaxPO)
+	}
+	// if value decreased and fell below the edge limit, then the freed capacity is added
+	if !ch.increased() {
+		s.rs.Capacity += half
+		// the chunks of this batch are unreserved in PO depth
+		return s.unreserve(b.ID, s.rs.Depth)
+	}
+	// otherwise value increased and we need to make capacity for its size
+	s.rs.Capacity -= half
+	// double means it got from no reserve to reserve upto full depth
+	if ch.double() {
+		s.rs.Capacity -= half
+	}
+	radius := s.rs.Depth - uint8(ch.new)
+	if err := s.unreserve(b.ID, radius); err != nil {
+		return err
+	}
+	return s.release(b)
```

Comment:
release is misleading. infers you might be releasing a locked resource. evict?

Reply:
in a way  yes, but evict is better thanks